plugins/inputs/turbostat/turbostat.go

```diff
@@ -0,0 +1,118 @@
+package turbostat
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"strconv"
+	"strings"
+
+	"github.com/influxdata/telegraf"
+)
+
+// Converts an arbitrary string to a snake case key.
+func cleanKey(s string) string {
+	words := make([]string, 0)
+	word := make([]rune, 0)
+	flush := func() {
+		if len(word) > 0 {
+			words = append(words, string(word))
+			word = make([]rune, 0)
+		}
+	}
+	for _, c := range s {
+		if c >= 'A' && c <= 'Z' {
+			c = c - ('A' - 'a')
+			word = append(word, c)
+		} else if (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9') {
+			word = append(word, c)
+		} else {
+			flush()
+			if c == '%' {
+				words = append(words, "percent")
+			}
+		}
+	}
+	flush()
+	return strings.Join(words, "_")
+}
+
+// Returns whether a string represents a tag or not.
+// Turbostat appears to only use integers and "-".
+func isTag(s string) bool {
+	if s == "-" {
+		return true
+	}
+	for _, c := range s {
+		if c < '0' || c > '9' {
+			return false
+		}
+	}
+	return true
+}
+
+// Converts keys and values to tags and fields.
+// Keys and values are easy to read from Turbostat stdout.
+// Tags and fields are easy to submit to Telegraf.
+func toTagsAndFields(keys, values []string) (map[string]string, map[string]any, error) {
+	if len(values) > len(keys) {
+		msg := "too many values: keys=%s, values=%s"
+		err := fmt.Errorf(msg, keys, values)
+		return nil, nil, err
+	}
+	tags := make(map[string]string, 0)
+	fields := make(map[string]any, 0)
+	for i := range values {
+		k := cleanKey(keys[i])
+		if k == "cpu" || k == "core" || k == "apic" || k == "x2apic" {
+			if !isTag(values[i]) {
+				return nil, nil, fmt.Errorf("invalid tag: %s", values[i])
+			}
+			tags[k] = values[i]
```

Comment:
It seems that not all `tags` are captured here. In addition to what you already have, you should add the missing strings from `BIC_GROUP_TOPOLOGY` (https://github.com/torvalds/linux/blob/f1a3944c860b0615d0513110d8cf62bb94adbb41/tools/power/x86/turbostat/turbostat.c#L278-L286), such as `package` and `die`, among others. Would be nice to have these new tags tested.
You can also extract tag checking to separate method.

Reply:
Done, although I am not 100% sure I have captured all the tags.

Reply:
@marcv81 After a quick analysis of the aforementioned `turbostat` code, it seems to me that `node` is missing.

Reply:
Fixed, thanks.